src/main/java/de/tum/in/www1/artemis/service/SubmissionService.java

```diff
@@ -531,4 +576,54 @@ public void checkIfExerciseDueDateIsReached(Exercise exercise) {
         return submissions;
     }
 
+    /**
+     * This method gets all complaints of an exercise and returns them together with their corresponding submission in a DTO
+     *
+     * @param exerciseId the exerciseId of the exercise of which the complaints are fetched
+     * @param principal the current user
+     * @param isAtLeastInstructor if the user is an instructor
+     * @return a list of DTOs containing a complaint and its submission
+     */
+    public List<SubmissionWithComplaintDTO> getSubmissionsWithComplaintsForExercise(Long exerciseId, Principal principal, boolean isAtLeastInstructor) {
+        List<SubmissionWithComplaintDTO> submissionWithComplaintDTOs = new ArrayList<>();
+
+        // get all complaints which belong to the exercise
+        List<Complaint> complaints = complaintRepository.getAllComplaintsByExerciseIdAndComplaintType(exerciseId, ComplaintType.COMPLAINT);
+        var complaintMap = complaints.stream().collect(Collectors.toMap(complaint -> complaint.getResult().getId(), value -> value));
+
+        if (complaints.isEmpty()) {
+            return submissionWithComplaintDTOs;
+        }
+        // get the ids of all results which have a complaint, and with those fetch all their submissions
+        List<Long> submissionIds = complaints.stream().map(complaint -> complaint.getResult().getSubmission().getId()).collect(toList());
+        List<Submission> submissions = submissionRepository.findBySubmissionIdsWithEagerResults(submissionIds);
+
+        // add each submission with its complaint to the DTO
+        submissions.forEach(submission -> {
+            Result resultWithComplaint = submission.getResultWithComplaint();
+            if (resultWithComplaint == null) {
+                return;
+            }
+            // get the complaint which belongs to the submission
+            Complaint complaintOfSubmission = complaintMap.get(resultWithComplaint.getId());
+            prepareComplaintAndSubmission(complaintOfSubmission, submission);
+            submissionWithComplaintDTOs.add(new SubmissionWithComplaintDTO(submission, complaintOfSubmission));
+        });
```

Comment:
How about utilizing `filter` here, just before the forEach, instead of returning from the forEach?
i.e.
```java
submissions
.map(submission → submission.getResultWithComplaint)
.filter(result → result != null)
.forEach(...)

Reply:
I guess its a bit better, I put a filter function before the forEach.